### DISCUSSION TITLE
How to add an attribute to all spans in a trace?

### DISCUSSION THREAD

**Discussion by kroehre** [open-telemetry/opentelemetry-ruby#1272] [Q&A] [answered]
Hi! I'm trying to understand how to add an attribute to all spans in a trace. As I understand it this is sometimes called "correlation context". I've tried using `OpenTelemetry::Baggage` following the example [here](https://github.com/open-telemetry/opentelemetry-ruby/discussions/701) but that doesn't seem to add attributes to spans. Thank you

**Comment by fbogsany** [comment, marked as answer]
Baggage does not automatically add attributes to spans in a trace. It can propagate "correlation context" but that's not its only use.

Typically, the easiest ways to add attributes to all spans are using a `SpanProcessor`, which receives `on_start` and `on_finish` callbacks: https://github.com/open-telemetry/opentelemetry-ruby/blob/113e5133afd7a42c30273d860330008c24530d6d/sdk/lib/opentelemetry/sdk/trace/span_processor.rb#L15-L33
or using a `Sampler`, which can return attributes to be added to a span alongside the sampling decision: https://github.com/open-telemetry/opentelemetry-ruby/blob/113e5133afd7a42c30273d860330008c24530d6d/sdk/lib/opentelemetry/sdk/trace/samplers.rb#L20-L39

**Reply by kroehre** [reply]
This works! Thank you.

Is `OpenTelemetry::Context#with_value` an appropriate way to assign correlation context attributes to be later assigned to child spans by the span processor?

**Reply by robbkidd** [reply]
I've [used](https://github.com/robbkidd/otel-ruby-rails/commit/f7e28dd35dfd72885811e5e664ee227dc87b3083) the `Baggage#set_value` API to set baggage attributes and use the returned `Context` in a call to `Context#with_current`. It's been a while since I experimented with the Context and Baggage interfaces, so I'm not positive this is the most ergonomic means to set baggage. Also in that commit is the definition of a custom BaggageSpanProcessor to duplicate baggage data onto spans and the wiring of it into the configuration. 

📓  Using Baggage and a span processor for this can only add attributes to spans _as they are started_. In other words, adding to baggage will not mutate the current span; if you want something on current and children, you'll need to add it to current _and_ to baggage.

📔  As far as I am aware, there is nothing in OpenTelemetry today that allows for pushing attributes up from child spans to parent spans. (There is no mechanism to mutate a parent context to have an effect on a parent span and, even if there were, [span processors `#on_finish` by spec does not have access to _any_ context and spans are immutable when passed in](https://opentelemetry.io/docs/reference/specification/trace/sdk/#onendspan).) Using Baggage and a span processor will not add attributes to all spans within a trace, only the _child_ spans from the current context.

⚠️  [Baggage propagation is on by default](https://opentelemetry.io/docs/reference/specification/sdk-environment-variables/#general-sdk-configuration). When using auto-instrumentation with the defaults, the data you add to baggage will appear in a header added to _all_ outgoing HTTP requests. Do not put sensitive data in Baggage.

**Reply by robbkidd** [reply]
Following up with [an example of using Baggage for this and some shenanigans with a custom, off-spec CarryOn](https://github.com/honeycombio/example-greeting-service/pull/125/files?w=1) and associated span processor.
